os/fs/smartfs/smartfs_utils.c

```diff
@@ -1268,6 +1283,224 @@ int smartfs_createentry(struct smartfs_mountpt_s *fs, uint16_t parentdirsector,
 	return ret;
 }
 
+#ifdef CONFIG_SMARTFS_USE_SECTOR_BUFFER
+/****************************************************************************
+ * Name: smartfs_find_firstinvalidentry
+ *
+ * Description: Finds an invalid entry in the parent directory that can be
+ *              used to create a new file entry without allocating a new
+ *              sector. The first one to be found is used.
+ *
+ ****************************************************************************/
+
+int smartfs_find_firstinvalidentry(struct smartfs_mountpt_s *fs, uint16_t parentdirsector, struct smartfs_entry_s *invalidentry)
+{
+	uint16_t ret;
+	struct smart_read_write_s readwrite;
+	uint16_t psector;
+	uint16_t nextsector;
+	uint16_t offset;
+	uint16_t entrysize;
+	struct smartfs_entry_header_s *entry;
+	struct smartfs_chain_header_s *chainheader;
+
+	psector = parentdirsector;
+	entrysize = sizeof(struct smartfs_entry_header_s) + fs->fs_llformat.namesize;
+
+	/* Return error message if no invalid entry is found */
+	ret = -ENOENT;
+
+	/* Scan through the entire parent sector chain to find an invalid entry */
+	while (psector != SMARTFS_ERASEDSTATE_16BIT) {
+		/* Read the next sector */
+
+		readwrite.logsector = psector;
+		readwrite.count = fs->fs_llformat.availbytes;
+		readwrite.offset = 0;
+		readwrite.buffer = (uint8_t *)fs->fs_rwbuffer;
+		ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
+		if (ret < 0) {
+			goto errout;
+		}
+
+		/* Get the next chained sector */
+		chainheader = (struct smartfs_chain_header_s *)fs->fs_rwbuffer;
+		nextsector = SMARTFS_NEXTSECTOR(chainheader);
+
+		/* Search for an invalid entry in this sector */
+		offset = sizeof(struct smartfs_chain_header_s);
+		entry = (struct smartfs_entry_header_s *)&fs->fs_rwbuffer[offset];
+
+		while ((offset + entrysize) < readwrite.count) {
+			/* Check if this entry is invalid */
+
+			if (entry->flags != SMARTFS_ERASEDSTATE_16BIT && !(ENTRY_VALID(entry))) {
```

Comment:
If entry is invalidated, we don't have to care about another flag in entry->flags, I think.
Because If it is invalid entry, we don't use it as a entry anymore.

Reply:
I added `entry->flags != SMARTFS_ERASEDSTATE_16BIT` to counter the case when find_firstinvalidentry() would come across an erased sector and consider an entry with flags = ERASEDSTATE is invalid and return that for replaceentry.
This works, but it is not logically correct.

However, now with updated logic, find_fistinvalidentry will save the first entry found with flags = ERASEDSTATE and return that as a free entry.

Reply:
Then,
```
if (!ENTRY_VALID(entry)) {
    if (IS_EMPTY_ENTRY(entry)) {
        saved_psector = psector;
	saved_doffset = offset;
    } else {
        invalidentry->dsector = psector;
	invalidentry->doffset = offset;
	ret = OK;
	goto errout;
    }
}
```
This doesn't work? can you please test?